### ISSUE TITLE
Secure AKS Baseline with CAPI/CAPZ

### ISSUE THREAD

**Issue by eedorenko** [kubernetes-sigs/cluster-api-provider-azure#1243]
> ⚠️ Cluster API Azure maintainers can ask to turn an issue-proposal into a [CAEP](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/YYYYMMDD-template.md) when necessary. This is to be expected for large changes that impact multiple components, breaking changes, or new large features.

***Goals***
1. Be able to provision a production ready Secure AKS baseline  cluster with CAPI/CAPZ 

**User Story**
In order to be in compliance with [AKS Secure Baseline Architecture](https://github.com/mspnp/aks-secure-baseline) an AKS cluster should have the following properties and configurations:

|Feature|[Installed by Flux](https://github.com/kaizentm/multicluster-gitops/tree/main/infra/base/secure-aks-baseline)|Possible with CAPI/CAPZ|
|-------|--------------|---------|
|Different Subnets (system nodes, user nodes, ingress subnet)| | No. [Not possible to specify different subnets for different agent pools](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/e6a7730ff546aa6c7885083188571f06f5492a32/azure/services/managedclusters/managedclusters.go#L184)|
|Azure CNI plugun| | Yes |
|AAD Pod Managed Identity| Yes| Yes|
|[Azure RBAC for K8s Auth](https://docs.microsoft.com/en-us/azure/aks/manage-azure-rbac)| | No. [EnableRBAC and AadProfile profile properties are not set](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/e6a7730ff546aa6c7885083188571f06f5492a32/azure/services/managedclusters/managedclusters.go#L123) |
|Azure Key Vault CSI secret provider| Yes| Yes|
|Traefik + TLS| Yes | Yes|
|Azure private link to KV and ACR| | No. Not possible to configure a VNet in the yaml
|K8s network policy| Yes | Yes |
|Azure K8s policies| | Yes |
|ACR| | No. How do I define attachment to a private acr in the yaml? |
|Flux V2|  | Yes
|Azure Monitor Prometheus Scraping| Yes| Yes |
|Kured| Yes| Yes |

That said, there are a few "NOs" in the "Possible with CAPI/CAPZ" column that currently prevent using CAPI/CAPZ for a [Secure AKS cluster](https://github.com/mspnp/aks-secure-baseline). None of these issues seem to be super complicated to implement, it just needs to be done.

**Comment by nprokopic** [comment]
I could tackle this list :) but maybe refactoring from https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/757 should be done first?

**Comment by eedorenko** [comment]
FYI @craiglpeters @jackfrancis

**Comment by alexeldeib** [comment]
Thanks for making this list! Nicely highlights some of the areas for us to work on.

for private link, is all that’s needed a custom VNET? We do have that logic, the VNET is in the ManagedControlPlaneSpec: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/7744b89fe0aae67ec3f282ffd957c1214dc33a52/exp/api/v1alpha4/azuremanagedcontrolplane_types.go#L42

I think we could potentially expand that to include subID and RG, to allow more flexibility with custom vnets in real deployments?

**Comment by alexeldeib** [comment]
private ACR would be a role binding to the kubelet identity used by the cluster. we're defaulting to System identities, but if we allow user-managed identities, we could "solve" this by asking users to provide a user-managed identity preconfigured with the desired RBAC to their ACR. There is no location to specify the ACR name AFAIK, kubelet's internal  provider will figure out you are pulling an ACR image and try to auth with the credentials we supply.

**Comment by dtzar** [comment]
@alexeldeib are any of these gaps on the planned backlog?

**Comment by alexeldeib** [comment]
#1418 and #1376 are the top issues for managed clusters and both in progress. I think we should resolve those before much more feature work. #1396 is also pretty straightforward and should be a reasonable follow up shortly after. 

With those done, I'd want to flesh out the model we map to the AKS API, addressing some of the issues here. @LochanRn is already working towards AAD support (https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1407), but needs #1418 and #1376 first. 

Different subnets should be pretty straightforward. We also need to add user-assigned managed identities for Kubelet. I think then the whole table will be filled out?

I don't have an explicit timeline for all of that. But like I mentioned, #1418 and #1376 are in-progress en route to #1407 (thank you again @LochanRn!). If no one beats me to it, I'll probably pick up subnets and user-assigned managed identities.

/remove-lifecycle stale

**Comment by alexeldeib** [comment]
system pools are very close to merging, I've done the work to remove AzureManagedCluster (albeit will need to rebase and probably won't merge that before v0.5.x).

I'm not sure this is a good first issue, but maybe a good 2nd/3rd issue 🙂 the missing pieces after #1520 merges would be subnets and managed identities. Happy to provide more guidance if someone wants to pick this up.

/help-wanted

**Comment by dtzar** [comment]
I think this issue is still relevant although I believe there should be a refreshed version of the issue based on the newest developments (i.e. Workload identity and existing support for some of the things out of the box with add-ons).  

@salaxander @CecileRobertMichon

**Comment by dtzar** [comment]
This is still relevant IF the "No's" in the table have not been completed/addressed.  If CAPZ supports all those, then it can be closed.  

IMO it would be good to have a `clusterctl` [template flavor for AKS ](https://github.com/kubernetes-sigs/cluster-api-provider-azure/tree/89b354a2ce91b8948190ce18ae488a96f35a1123/templates/flavors) which already implements all or most of these things.

**Comment by dtzar** [comment]
@jackfrancis - do you know if the three NO's in the column are still relevant?

**Comment by jackfrancis** [comment]
@dtzar RBAC and private link feature is done as far as I can tell.

I know some iterative work has been done w/ custom VNET + subnet functionality. @mtougeron @dthorsen @luthermonson can you remind me if there's anything w/ respect to "different subnet per pool" that's not yet done?

Based on this description I would conclude that ACR integration is not complete:

- https://learn.microsoft.com/en-us/azure/aks/cluster-container-registry-integration

@CecileRobertMichon are you aware if any work has been done on that? (also feel free to clarify anything incorrect above)

**Comment by CecileRobertMichon** [comment]
Per-pool subnet work is this unmerged PR: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2695 cc @LochanRn 

I don't know that any work has been done to integrate ACR but if that's exposed through the AKS APIs should be straightforward to add. Probably a good idea to track that one in a separate issue.

**Comment by mtougeron** [comment]
[Virtual Network service endpoints](https://learn.microsoft.com/en-us/azure/virtual-network/virtual-network-service-endpoints-overview) is now possible but creating a specific [Private Link](https://learn.microsoft.com/en-us/azure/private-link/private-link-overview) to ACR (as defined by [Connect privately to an Azure container registry using Azure Private Link](https://learn.microsoft.com/en-us/azure/container-registry/container-registry-private-link)) is not possible yet. I've been talking with a coworker about implementing private links; it's definitely something we can take on.

**Comment by dtzar** [comment]
Great!  I'd recommend closing this issue out now having filed the remaining gap as an issue:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2897

And new feature request for the AKS Baseline: https://github.com/mspnp/aks-baseline/issues/359
